cpputil/unstructured/include/gridtools/next/mesh.hpp

```diff
@@ -24,19 +44,26 @@ namespace gridtools {
             }
 
             template <class Connectivity>
-            GT_FUNCTION auto skip_value(Connectivity const &connectivity) {
+            GT_FUNCTION auto skip_value(Connectivity const &connectivity)
+                -> decltype(connectivity_skip_value(connectivity)) {
                 return connectivity_skip_value(connectivity);
             }
         } // namespace connectivity
 
         namespace mesh {
-            template <typename Key, typename Mesh>
-            decltype(auto) mesh_connectivity(Mesh const &mesh);
+            // models hymaps as Mesh
+            // TODO(havogt): protection: only if values model the Connectivity concept and keys are meta::list<> of
+            // locations
+            // TODO probably remove?
+            template <typename Key, typename Hymap> // TODO protect, only hymaps allowed
+            decltype(auto) mesh_connectivity(Key const &, const Hymap &mesh) {
```

Comment:
The fallbacks that defined that way are fragile. it can happen that the user do have its own overload but your overload somehow win. To protect that you should use the same technique as in the sid concept (abuse the fact the the worst overload ever is f(...)): 

```
struct not_provided;
not_provided mesh_connectivity(...);

template <class Key,
                  class Mesh,
                  class K = meta::rename<meta::list, Key>,
                  class Res = decltype(mesh_connectivity(K(), std::declval<Mesh const&>())),
                  std::enable_if_t<std::is_same<Res, not_provided>::value, int> = 0>
auto connectivity(Mesh const& mesh) -> decltype(at_key<K>(mesh)) {
  return at_key<K>(mesh);
}

template <class Key,
                  class Mesh,
                  class K = meta::rename<meta::list, Key>,
                  class Res = decltype(mesh_connectivity(K(), std::declval<Mesh const&>())),
                  std::enable_if_t<!std::is_same<Res, not_provided>::value, int> = 0>
Res connectivity(Mesh const& mesh) {
  return mesh_connectivity(K(), mesh);
}

```